ethers-core/src/types/crypto/signature.rs

```diff
@@ -92,22 +99,27 @@ impl Signature {
             RecoveryMessage::Data(ref message) => hash_message(message),
             RecoveryMessage::Hash(hash) => hash,
         };
-        let message = Message::parse_slice(message_hash.as_bytes())?;
 
-        let (signature, recovery_id) = self.as_signature()?;
-        let public_key = Secp256k1::recover(&message, &signature, &recovery_id)?;
+        let (recoverable_sig, _recovery_id) = self.as_signature()?;
+        let verify_key = recoverable_sig.recover_verify_key(message_hash.as_bytes())?;
 
-        Ok(PublicKey::from(public_key).into())
+        let uncompressed_pub_key = K256PublicKey::from(&verify_key).decompress();
+        if uncompressed_pub_key.is_some().into() {
+            let pub_key: K256PublicKey = K256PublicKey::from(uncompressed_pub_key.unwrap());
+            Ok(PublicKey::from(pub_key).into())
+        } else {
+            Err(SignatureError::RecoveryError)
+        }
     }
 
     /// Retrieves the recovery signature.
     fn as_signature(&self) -> Result<(RecoverableSignature, RecoveryId), SignatureError> {
         let recovery_id = self.recovery_id()?;
         let signature = {
-            let mut sig = [0u8; 64];
-            sig[..32].copy_from_slice(self.r.as_bytes());
-            sig[32..].copy_from_slice(self.s.as_bytes());
-            RecoverableSignature::parse(&sig)
+            let gar: &GenericArray<u8, U32> = GenericArray::from_slice(self.r.as_bytes());
+            let gas: &GenericArray<u8, U32> = GenericArray::from_slice(self.s.as_bytes());
+            let sig = K256Signature::from_scalars(*gar, *gas)?;
```

Comment:
FYI, I believe you can avoid the `GenericArray` crap here by doing:

```suggestion
            let sig = K256Signature::from_scalars(self.r.as_fixed_bytes().into(), self.r_as_fixed_bytes().into())?;
```

or if `into()` doesn't work, try `as_ref()`